https://github.com/parse-community/parse-dashboard/issues/151

fn-faisal opened the pull request:
[Issue 151 link](https://github.com/parse-community/parse-dashboard/issues/151)

- The resolution implemented in this PR for this issue is to give users access to an option in the menu where they can change the default key for a given class. 
- Throughout the Dashboard, anywhere that there is a pointer to this class object, the specified key's value (chosen in the edit menu option -> change pointer key) is used instead of objectId.

If no key is specified, the pointer value falls back to objectId.
<img width="757" alt="Screenshot 2021-06-03 at 6 31 22 PM" src="https://user-images.githubusercontent.com/23558802/120653334-06060e80-c49a-11eb-904b-754530f7fff4.png">
<img width="757" alt="Screenshot 2021-06-03 at 6 31 31 PM" src="https://user-images.githubusercontent.com/23558802/120653361-0acac280-c49a-11eb-99c5-c28bbf119f26.png">
<img width="757" alt="Screenshot 2021-06-03 at 6 31 41 PM" src="https://user-images.githubusercontent.com/23558802/120653371-0bfbef80-c49a-11eb-81ec-1ea26e11b051.png">
<img width="634" alt="Screenshot 2021-06-03 at 6 32 00 PM" src="https://user-images.githubusercontent.com/23558802/120653379-0e5e4980-c49a-11eb-852a-fcd70e8fbd24.png">

fn-faisal commented:
I will revert the package-lock.json but I'd need to add sweet alert dependency.

mtrezza commented:
Can you please rename the title accordingly?

fn-faisal commented:
> Thanks for this PR.
> 
> Can you please rename the PR title accordingly?

I have updated the PR title.

mtrezza commented:
Thank you.

There are some issues that come to mind:
- Does each cell trigger a separate query to fetch the representation, or is that done together with the paging mechanism as a bulk-fetch?
- Where is the information about the pointer key stored? Is it sustainable enough?

fn-faisal commented:
> Thank you.
> 
> * Does each cell trigger a separate query to fetch the representation, or is that done together with the paging mechanism as a bulk-fetch?
> * Where is the information about the pointer key stored? Is it sustainable enough?
> * Do we need the SweetAlert component? We already have existing components to achieve the same.

- For each cell, if the type is pointer and the pointer key is not objectId, it fetches the object using the get method of the Parse Query object.
- To store pointer data, localStorage is used to store Class name as key and the key as value.
- A similar PR was made using b4a dash and it has sweet alert so I went with it and this PR is based off of that one, but I can revert this to Parse Dashboard specific alter dialogue.

fn-faisal commented:
I've pushed a commit that will optimize the query. The second query has been removed and instead I've added include and select on the first query that fetches the data. I'll update the dialog in the next commit.

fn-faisal commented:
<img width="634" alt="Screenshot 2021-06-05 at 12 03 27 AM" src="https://user-images.githubusercontent.com/23558802/120850885-7c396c80-c591-11eb-86a1-30569e9f8b38.png">

The dialog has been updated and the sweet alert2 dependencies have been removed.

fn-faisal commented:
<img width="634" alt="Screenshot 2021-06-05 at 12 08 52 AM" src="https://user-images.githubusercontent.com/23558802/120851566-6f694880-c592-11eb-962b-f5fd91a3fee1.png">

The dialogue to change pointer value has been updated. The sweet alert2 deps have been removed.

mtrezza commented:
Great, kindly let me know when this is ready for review.

fn-faisal commented:
I've cleaned up the code and the PR is ready to be reviewed.

mtrezza commented on src/dashboard/Data/Browser/PointerKeyDialog.react.js:
```diff
@@ -0,0 +1,72 @@
+        type={Modal.Types.INFO}
+        icon='warn-outline'
+        title={'Change pointer value for class '+this.props.className}
+        subtitle={hasColumns ? 'The column will be used throughout dashboard as pointer value for this class.' : 'There are no columns on this class that can be set a pointer value.'}
```
Fix layout issue where text overlays icon:

<img width="546" alt="image" src="https://user-images.githubusercontent.com/5673677/120922034-3ed6fb00-c6c7-11eb-9a27-dddffe761aae.png">

fn-faisal replied:
<img width="634" alt="Screenshot 2021-06-07 at 5 55 17 PM" src="https://user-images.githubusercontent.com/23558802/121021496-e3863500-c7ba-11eb-955f-9c5180789856.png">

mtrezza commented on src/dashboard/Data/Browser/PointerKeyDialog.react.js:
```diff
@@ -0,0 +1,72 @@
+        type={Modal.Types.INFO}
+        icon='warn-outline'
+        title={'Change pointer value for class '+this.props.className}
+        subtitle={hasColumns ? 'The column will be used throughout dashboard as pointer value for this class.' : 'There are no columns on this class that can be set a pointer value.'}
```
Rename `pointer value` to `pointer key`. We should use the same term throughout code and UI.

mtrezza commented on src/dashboard/Data/Browser/PointerKeyDialog.react.js:
```diff
@@ -0,0 +1,72 @@
+        icon='warn-outline'
+        title={'Change pointer value for class '+this.props.className}
+        subtitle={hasColumns ? 'The column will be used throughout dashboard as pointer value for this class.' : 'There are no columns on this class that can be set a pointer value.'}
+        confirmText='Update pointer value'
```
Rename `pointer value` to `pointer key`. We should use the same term throughout code and UI.

mtrezza replied:
This is marked as resolved, but the change has not been made.

mtrezza commented on src/dashboard/Data/Browser/PointerKeyDialog.react.js:
```diff
@@ -0,0 +1,72 @@
+      <Modal
+        type={Modal.Types.INFO}
+        icon='warn-outline'
+        title={'Change pointer value for class '+this.props.className}
```
Remove class name from dialog, due to possible layout issues with large class names; in stead use:

> Change pointer key for this class

mtrezza commented on src/components/BrowserCell/BrowserCell.react.js:
```diff
@@ -235,18 +234,38 @@ export default class BrowserCell extends Component {
+      let dataValue = value.id;
+      if ( defaultPointerKey !== 'objectId' ) {
+        dataValue = value.get(defaultPointerKey);
+        if ( dataValue && typeof dataValue === 'object' ){
```
We need something else than `Invalid Value` in case the pointer key is set to a field that is `(undefined)`.

<img width="154" alt="image" src="https://user-images.githubusercontent.com/5673677/120922603-654a6580-c6ca-11eb-8039-a24371b10a0e.png">

I suggest to display `(undefined)` which is the true current value of the pointer key, or maybe you have a different idea.

fn-faisal replied:
I can replace the square brace but from my understanding of the dashboard, the undefined is not clickable. The idea that I had while implementing this was that if there's an issue while loading the pointer key value, display something but still leave the field clickable to allow users to navigate the between the pointer relation.
While undefined indicates that there is no or the row doesn't point to any row in the respective table.

mtrezza replied:
The word "invalid" has a different meaning than "undefined". If a pointer key value is "undefined", we should convey that it is simply "undefined". But we also cannot display it the same as `(undefined)`, which means that no value is set for the field.

Since we will abandon the pill design with https://github.com/parse-community/parse-dashboard/pull/1708 I think we end up with these ambiguations:

- a) `(undefined)` = no value is set for field
- b) `undefined ↗️` = pointer field value is defined as string "undefined"
- c) `(undefined) ↗️` = pointer field value is not defined

This means that a pointer key value that is undefined can only be displayed as (c).

fn-faisal replied:
Ok, I'll update it to match the PR branch #1708 and update this PR to match (c)

mtrezza replied:
I think we should merge the pointer redesign PR first, then merge master into this PR?

fn-faisal replied:
I have also updated the PR with the new pointer design.

fn-faisal commented:
I have made the changes and code is ready to be reviewed again.

mtrezza commented:
Thanks, can you please take a look at https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r655729128

mtrezza commented:
@fn-faisal It would be great if we could merge this amazing PR. Would you mind merging this with master to resolve any conflicts? And do you think we should merge the pointer redesign PR first, and merge this PR afterwards?

fn-faisal commented:
@mtrezza I think it would we should merge the pointer PR first before merging this PR. I haven't yet caught this PR with the pointer PR yet.

mtrezza commented:
Ok, I will ping you here once the other PR has been merged.

mtrezza commented:
@fn-faisal #1708 has been merged. Could you rebase this PR on master to merge it next?

fn-faisal commented:
@mtrezza I updated the PR but running it locally, I encountered bugs that I updated for Pill when displaying array of pointers etc. The PR is ready for review.

mtrezza commented:
> I encountered bugs that I updated for Pill when displaying array of pointers and cell section/edit value

Do you mean you fixed bugs that were not part of this PR? If so, can you point out which changes these are?

fn-faisal commented:
@mtrezza These were new issues.

1. The cell border that appears when clicking on a cell wasn't appearing but the value was copyable.
2. The data from the edit field that appears, when double clicking on a cell and entering a new value and then clicking away to save, wasn't appearing in the cell.
3. The array of pointers was not rendering properly with the new pill design.

mtrezza commented:
Normally, bugs that are not part of the PR would require a separate PR. But I think in this case we can make an exception and reduce the work because we have a couple of PRs to merge that make changes to the same files and already have undergone multiple rebasing.

fn-faisal commented:
@mtrezza I encountered the bugs while resolving merge conflicts. They were not present in the master branch.

mtrezza commented:
Got it, in that case they can be considered part of this PR. Thanks for clarifying, and for the attention in finding these bugs. I will take a look later if I find any other side effects.

mtrezza commented:
Could you just rebase on the main branch? Apologies for all the rebasing, we just have so many great PRs that are being merged.

fn-faisal commented:
> Could you just rebase on the main branch? Apologies for all the rebasing, we just have so many great PRs that are being merged. This PR will be merged next if all goes well.

I have resolved the merge conflicts for this PR.

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -1432,7 +1489,17 @@ class Browser extends DashboardView {
       }
     }
     let extras = null;
-    if (this.state.showCreateClassDialog) {
```
Is this change intended?

mtrezza commented on src/components/BrowserCell/BrowserCell.react.js:
```diff
@@ -23,12 +23,149 @@ export default class BrowserCell extends Component {
+        content = <span>&nbsp;</span>;
+        classes.push(styles.empty);
+      } else if (this.props.type === 'Pointer') {
+        const defaultPointerKey = await localStorage.getItem(this.props.value.className) || 'objectId';
```
I assume there is currently no caching mechanism, correct? So if the same pointer is used with the same objectId in multiple rows, there will be a DB request for each cell? I would not make this is a requirement for this PR, but I think that would be one of the possible improvements we could mention in the top post of this PR, so people are aware of this caveat.

fn-faisal replied:
Yes, it'll perform a DB operation for each row. I'll mention it on the PR.

mtrezza commented on src/components/BrowserCell/BrowserCell.react.js:
```diff
@@ -329,16 +480,32 @@ export default class BrowserCell extends Component {
       classes.push(styles.required);
     }
 
+    // console.log(value, this.state.value);
```
Can this be removed?

fn-faisal replied:
I have removed the console.log

mtrezza commented on src/components/BrowserCell/BrowserCell.react.js:
```diff
@@ -237,20 +373,36 @@ export default class BrowserCell extends Component {
+      if( defaultPointerKey !== 'objectId' ) {
+        dataValue = value.get(defaultPointerKey);
+        if ( dataValue && typeof dataValue === 'object' ){
+          if ( dataValue instanceof Date ) {
```
This looks like strange formatting with the space inside the brackets. Do we run lint over this usually?

mtrezza commented:
Can you please also add a shot feature documentation including any [limitations](https://github.com/parse-community/parse-dashboard/pull/1723/files#r688584766)? We have a new section in README.md:

https://github.com/parse-community/parse-dashboard/blob/master/README.md#features

mtrezza commented:
I assume your [comment](https://github.com/parse-community/parse-dashboard/pull/1754#issuecomment-899504959) was for this PR?

mtrezza commented:
@fn-faisal I slightly changed the feature description, can you please look over it whether I got it right?

fn-faisal commented:
@mtrezza looks good to me 👍

mtrezza commented:
Actually I just realized that the example in the description doesn't work, it's not possible to use the `email` field of user as pointer key. Why is that?

Another question, when changing the pointer key back to `objectId`, does that remove any custom pointer key setting or does it still trigger server requests to resolve the poiner key?

fn-faisal commented:
@mtrezza The default system keys aren't passed to the pointer keys dialog. I added the objectId manually. If the email is a key not created by the user, it may be getting excluded from the columns passed to the pointer keys dialog.

Changing the key back to pointer key just sets the pointer key in localStorage back to objectId and I think the browser cell does not fetch for objectId.

mtrezza commented:
> I think the browser cell does not fetch for objectId.

Can you change this to make sure there is no fetching? If it still fetches that would be a bug. Maybe just completely delete the pointer key entry from the browser storage.

fn-faisal commented:
@mtrezza from the below screen shots, it looks like there's not fetch for pointer key. The pointer key includes the Parse object on which the value is picked from using the `.get` method. In the image below, the pointer key is not objectId and there's no extra network call.
<img width="1014" alt="Screenshot 2021-08-16 at 9 55 22 PM" src="https://user-images.githubusercontent.com/23558802/129601093-0bad5639-2dfb-4a78-8a1c-9adb96c3f4ab.png">

mtrezza commented:
Yes, I can see there is a condition in code that doesn't fetch if the the pointer key is set to `objectId`, so we should be good.

mtrezza commented on src/components/BrowserCell/BrowserCell.react.js:
```diff
@@ -23,12 +23,149 @@ export default class BrowserCell extends Component {
   }
 
-  componentDidUpdate(prevProps) {
+  async componentDidUpdate(prevProps) {
```
What are these changes? They seem not to be part of this PR?

fn-faisal replied:
The value of pointer key column is fetched from local storage using await, that's why async is added here.
https://github.com/fn-faisal/parse-dashboard/blob/feat/151-pcd/src/components/BrowserCell/BrowserCell.react.js#L376

mtrezza replied:
I actually mean the many changes in this method:

https://github.com/parse-community/parse-dashboard/pull/1723/files?file-filters%5B%5D=.js#diff-4e4d2ae32c4431bc66b52e869432797428508c6a3cb0811018e658525f0637c9R33-R168

fn-faisal replied:
I moved the changes from the component's render method to the did update method. I was having issues with async loading of data ( like key from local storage ).

mtrezza replied:
When you say "moved", I don't see nearly the same amount of lines removed from the `render` method. That looks strange to me.

fn-faisal replied:
Oh, sorry, I copied cell rendering logic to re-render on component updates for async calls. But this is for component updates. The lines in the render are for the initial render.

mtrezza replied:
Do you mean that you copied these lines from somewhere else? From where?

fn-faisal replied:
The lines are what render the browsercell based on the type of data passed. When the key is fetched from localstorage, it fetches and renders the pointer key's value. These lines are also present in the render function but since we need async calls, the lines are also partially copied to the componentDidUpdated method.
Previously, the component wasn't using states but I store the cell data, once fetched from pointer key, to state upon the resolution of promise callbacks and the component's DidMount function updates the cell value accordingly.

mtrezza replied:
> These lines are also present in the render function 

Can you post a link to that render function that contains these lines?

fn-faisal replied:
Before it used to simply get the cell values from props and no async calls were place.
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L217

I updated it to store cell values in state so when the pointer key column name is fetched from localstorage and the value ,for this column, is fetched using js SDK .get method, the component state is updated with the value.
https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r690610276

The lines in the render function that check the data type and show how the cell should be rendered are present in render function (I wanted to clean it up too by moving the lines to a separate function and call it in the didMount and didUpdate component lifecycle function function but didn't get around to it)

These lines check the cell data type and how it should be rendered (I needed to re-render the cell data once it's fetched from the localStorage+SDK so I had to copy them to didUpdate function for when the cell data in the component state updates).
This is where the logic of how the cell should renders starts.
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L222

and ends around this line
https://github.com/parse-community/parse-dashboard/blob/1ddf29e84404ed36db86fc496fab82145add8bb1/src/components/BrowserCell/BrowserCell.react.js#L330

mtrezza replied:
> I wanted to clean it up too by moving the lines to a separate function and call it in the didMount and didUpdate component lifecycle function function but didn't get around to it

Thanks for explaining this in detail, I understand now. I think it would be really great if you could do the clean-up before we merge this PR. Duplicate code is often difficult to maintain, if someone later makes a change only in 1 part of code, but not in the other.

mtrezza commented:
@fn-faisal Please feel free to request a review if you think this is ready

mtrezza commented:
@fn-faisal Could you please resolve the conflicts for review?

mtrezza commented:
This is almost ready, the only thing I found was an issue if the pointer key is set to another pointer column or a file column.


https://user-images.githubusercontent.com/5673677/132421198-99971e1f-abec-4040-8d95-7bc335a2a2a8.mov

mtrezza commented:
Does this work on a per-app basis?

I am thinking, if I have multiple tabs of different dashboards open in the same browser session, would the pointer column settings influence each other?

fn-faisal commented:
@mtrezza I will update it to work for per app basis and display ID if the shadow value is undefined.

fn-faisal commented:
@mtrezza I had a suggestion about the pointer cell. What if an icon is displayed for the column cell that uses objectId and icon isn't shown for shadow values. 

<img width="986" alt="Screenshot 2021-09-13 at 7 11 21 PM" src="https://user-images.githubusercontent.com/23558802/133099359-d4d239a3-c555-4f29-b94f-64010829fd38.png">

fn-faisal commented:
@mtrezza I have added appId along with className to save pointerKey + if the shadow value is null, now the cell displays pointer value instead.

mtrezza commented on Parse-Dashboard/public/index.html:
```diff
@@ -0,0 +1,40 @@
+<body>
+<div id="browser_mount"></div>
+<script type="text/javascript">
+    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
```
Can you please remove this?

mtrezza commented on .vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+{
```
Can you please remove this and instead add this to `.gitignore`:

```
# visual studio code
.vscode
```

mtrezza commented on package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "parse-dashboard",
-  "version": "3.0.0",
+  "version": "2.2.0",
```
Can you please merge master into this branch? This should be based on master.

fn-faisal commented:
@mtrezza I have removed `.vscode`, updated `.gitignore` and  reverted `package-lock.json`

mtrezza commented on CHANGELOG.md:
```diff
@@ -15,6 +15,10 @@ Details:
 - Purpose: official release
 - Suitable environment: production
 
+## New Features
```
Can you remove any change to the changelog? changelog entry will be generated automatically from now on based on the pull request title

fn-faisal replied:
I have removed the entry from CHANGELOG.md file

mtrezza commented on src/components/Button/Button.example.js:
```diff
@@ -10,6 +10,7 @@ import Button from 'components/Button/Button.react';
 
 export const component = Button;
 
+
```
can you remove this newline?

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -657,6 +684,30 @@ class Browser extends DashboardView {
       query.ascending(field)
     }
 
+    if (field !== 'objectId') {
```
This seems to be a merge issue; this code does not exist in the master branch anymore, see https://github.com/parse-community/parse-dashboard/pull/1800.

I am worried that there is some code in this PR that is obsolete but is still here due to master merging. Can you please take a minute and look through tall the change in this PR, compare it to the current master, and see if all the changes are related to your PR? Sorry for the extra work, but we better take an extra look before we introduce a bug.

mtrezza commented:
It seems there is an issue in the following case, given that:

- the `User` class has field `installation` that is a pointer to the `Installation` class
- the `Installation` class has field `user` that is a pointer to the `User ` class

Steps to reproduce:
1. In `User` class set pointer key to `installation` field
2. In `Installation` class refresh data table
3. Note that In `Installation` class the field values in column `user` show `(undefined)`, but should show the installation's `objectId`.

fn-faisal commented:
@mtrezza For the first issue regarding the pointers, I tried in the following video but could not reproduce the issue.

Uploading Screen Recording 2021-09-29 at 8.22.15 PM.mov…

fn-faisal commented:
@mtrezza For the first issue regarding the pointers, I tried the steps above in the following video but could not reproduce the issue.

https://user-images.githubusercontent.com/23558802/135311902-e1e7249d-f1f3-4e0a-820c-e6d5e4295328.mov

mtrezza commented:
Thanks, I'll take another look. Maybe you can take a look [this](https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r715015482) meanwhile.

mtrezza commented:
To make sure it doesn't get overlooked, the issue review is on hold because of https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r715015482

fn-faisal commented:
@mtrezza I've reverted the changes for comment [#1723 (comment)](https://github.com/parse-community/parse-dashboard/pull/1723#discussion_r715015482)

mtrezza commented:
Great, did you look through the whole PR if there are any other modifications that are not part of this PR?

fn-faisal commented:
@mtrezza I went through the PR and it only seems to have changes related to pointer key issue.

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -644,6 +648,29 @@ class Browser extends DashboardView {
     }
   }
 
+  setRequiredColumnFields() {
```
Is this also part of this PR?

fn-faisal replied:
I have removed this function. It was not being used anywhere and may have got in during a merge conflict resolution.

mtrezza commented:
🎉 This pull request has been released in version [3.3.0-alpha.2](https://github.com/parse-community/parse-dashboard/releases/tag/3.3.0-alpha.2)
